src/MongoDB.Bson/IO/JsonReader.cs

```diff
@@ -956,19 +956,112 @@ private BsonValue ParseBinDataExtendedJson()
         {
             VerifyToken(":");
 
-            var bytesToken = PopToken();
-            if (bytesToken.Type != JsonTokenType.String)
+            byte[] bytes;
+            BsonBinarySubType subType;
+
+            var nextToken = PopToken();
+            if (nextToken.Type == JsonTokenType.BeginObject)
             {
-                var message = string.Format("JSON reader expected a string but found '{0}'.", bytesToken.Lexeme);
+                ParseBinDataExtendedJsonCanonical(out bytes, out subType);
+            }
+            else
+            {
+                ParseBinDataExtendedJsonLegacy(nextToken, out bytes, out subType);
+            }
```

Comment:
The above lines parse either the canonical or legacy formats of $binary extended JSON forms.

The remaining lines in this method are common to both cases.